Add least_connections load balancing algorithm.

alowde opened the pull request:
This commit adds a new option to the load-balance setting, 'least_connections'.
In this mode the controller keeps a count of connections to each upstream and
routes incoming connections to the upstream with the least connections open.
This is recommended for use with evenly-resourced upstream servers when
requests have a broad distribution in time to process, for example if some
requests require the upstream server to make a connection to a slow external
service.

If all requests take a fairly similar time to process or the upstreams serve at 
different speeds then ewma or round_robin are likely more appropriate.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] CVE Report (Scanner found CVE and adding report)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation only

## How Has This Been Tested?
This PR includes Lua unit tests and end-to-end tests written using the ginkgo framework. 
The framework has been extended somewhat to support end-to-end tests that require
multiple upstreams with different configurations. All tests have been run including
existing tests, and all tests have passed.

## Checklist:
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/kubernetes/ingress-nginx/blob/main/CONTRIBUTING.md) guide
- [x] I have added unit and/or e2e tests to cover my changes.
- [x] All new and existing tests passed.
- [x] Added Release Notes.

## Does my pull request need a release note?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Added a new load-balancing algorithm 'least_connections'. If the load-balance setting in the config map is set to 'least_connections' the controller will attempt to route incoming connections to one of the upstream endpoints with fewest current open connections.
```

adkafka commented on rootfs/etc/nginx/lua/balancer/leastconn.lua:
```diff
@@ -0,0 +1,115 @@
+    self.peers = backend.endpoints
+
+    for _, endpoint_string in ipairs(normalized_endpoints_removed) do
+        ngx.shared.balancer_leastconn:delete(endpoint_string)
```
Removing these endpoints here may not have the desired affect. If a pod has a bunch of connections, briefly fails its readiness probe (causing it to no longer be included as a backend), we will remove it from being tracked here. But if the pod becomes ready again, it may still be handling a bunch of connections, but we will initialize its value to 0 again.

Maybe this is OK, but it is worth calling out.

alowde replied:
Yeah, that's not ideal behaviour. I think it would be safe to instead set expiration time to (say) 30 minutes. This would ensure we still track connection count for connections up to that time while not allowing the shared dict to grow indefinitely.

adkafka replied:
Setting it to expire asynchronously makes some sense, though it will be difficult to choose the TTL.

It makes sense that this least conn load balancer would still track pods that are handling connections, even if the pod is not ready. It should decrement the number of connections on a backend that finishes a request, even if that pod is not ready for example.

Is there a way we can tell the difference between a pod that is not ready now (but therefore may become ready soon) versus a pod that is being killed (and therefore will never become ready again, and would be safe to fully delete)?

alowde replied:
> Is there a way we can tell the difference between a pod that is not ready now (but therefore may become ready soon) versus a pod that is being killed (and therefore will never become ready again, and would be safe to fully delete)?

Not to the best of my knowledge, and that's the heart of the issue - all we know is that an upstream address has disappeared from the list of endpoints for that service. We could find out more by querying the API server but that introduces substantially more complexity that I think the ingress-nginx team would not want to take on.

adkafka replied:
This thread is still a bit concerning to me. It is unintuitive to me that a pod that becomes unready will result in ingress-nginx thinking it has 0 connections. Maybe setting the TTL to 2 days and just letting the entry in the map expire if we don't see it again in 2 days makes sense? (2 days is chosen somewhat arbitrarily)

ShadowJonathan replied:
I've noticed some oddities here, while working with pods which had its readiness checks interrupted (due to it asking the same port/endpoint as the incoming requests), causing it to receive extra connections for a while

See the below graphs
<img width="1483" alt="image" src="https://github.com/kubernetes/ingress-nginx/assets/22740616/e20105b2-fc64-43ed-b716-ef62748de3b2">

ShadowJonathan replied:
Some more correlation to this effect:
<img width="1486" alt="image" src="https://github.com/kubernetes/ingress-nginx/assets/22740616/b333c1e0-b656-4937-bed9-15a67d89936b">

ShadowJonathan replied:
Update: all of these effects were due to removing nodes from the table when they "go missing", plus there was an off-by-one error that has now been fixed on my fork, see [the other thread](https://github.com/kubernetes/ingress-nginx/pull/9025#discussion_r1421357694)

adkafka reviewed (commented):
Hey @alowde, I came across your PR because I too would like the least_connections load balancing algorithm for nginx. Thanks for making this!

I added some comments, but I'm no expert in lua load balancing algorithms. Hope it helps though. Would be great to get a least_connections algorithm added in.

Another worry I have with this implementation though, is we may end up choosing the same host over and over. I wonder if we could to least connection but by bucket. Ie, split the peers into N buckets (N=10?), and choose a random host within the bucket. This may be out of scope of this initial implementation, but I thought it was worth sharing. I wonder how other least_connection algorithms handle this (if at all).

alowde commented on rootfs/etc/nginx/lua/balancer/leastconn.lua:
```diff
@@ -0,0 +1,115 @@
+    local feasible_endpoints = {}
+
+    if #peers ~= 1 then
+        local lowestconns = 9999
```
I'm not certain what max int is in this case due to the various ways Lua could be compiled, but it's definitely higher than 9999 :). As a safe balance I've changed it to maximum signed 32 bit int, which should still be more connections than we're likely to support in other places.

adkafka replied:
This thread is marked as resolved, but I still see this set to `9999`. Once we have a pod w/ more than 10,000 connections, this will behave unexpectedly, right?

alowde commented on rootfs/etc/nginx/lua/balancer/leastconn.lua:
```diff
@@ -0,0 +1,115 @@
+        for _, peer in pairs(peers) do
+            local conns = endpoints:get(get_upstream_name(peer))
+            if conns == nil then
+              endpoints:set(get_upstream_name(peer),0,600)
```
My concern with depending on the sync method to handle garbage collection is trusting that the sync method will be reliably called with updates, and that we won't lose the list of peers between updates. With that said the main reason I was thinking to distrust sync was in case of nginx restarts - but then in that case the shared dict would be reset anyway.
I think we should set the expiration time to infinite.

adkafka replied:
I also agree the expiration time should be infinite. Do you have a reason to be suspicious of the sync method not being called reliably? I hope it is something we can depend on.

alowde replied:
No specific reason to be suspicious, call it a generalised sense of professional paranoia :). Will fix the expiration time shortly.

alowde commented on rootfs/etc/nginx/lua/balancer/leastconn.lua:
```diff
@@ -0,0 +1,115 @@
+        end
+
+        -- get peers with lowest connections
+        for _, peer in pairs(peers) do
```
Done

alowde commented:
Hey @adkafka, thanks very much for taking the time to review this. I've added commits addressing some of the issues and would be interested to discuss the rest further.

> Another worry I have with this implementation though, is we may end up choosing the same host over and over. I wonder if we could to least connection but by bucket. Ie, split the peers into N buckets (N=10?), and choose a random host within the bucket.

If there's multiple hosts with equal lowest connection count then this implementation will select one at random. If one host is significantly faster than the others it will get repeated requests, but the slower hosts will still get new requests as they complete old ones.

Originally my thinking was that if one server is handling most of the requests while others are slow, then the fast server should receive the requests. Your comment prompted me though to think about cases where one server is rapidly responding to requests with an error response (because of misconfiguration, error condition, or similar) and others are taking more time but succeeding. In this case we would actually produce a much worse result unless and until a health check kicked in and removed the bad peer.

I can think of a few options for handling this scenario and I'd be interested to hear thoughts from anyone:
1. Ignore it, and document the issue. 
2. Randomly select from n servers with the lowest connection count, where n is adjustable or adaptive (i.e. bucketing with just the lowest bucket considered). Partly mitigates the 'one bad peer' effect at the cost of less predictable behaviour.
3. Add an option for minimum connection count, effectively only balancing based on minimum connections above x. This would mean that peers with less than x open requests would be considered equal, which would reduce the impact of a peer returning quickly with a bad response.
  - Doesn't actually resolve the issue, just dilutes it at the cost of extra complexity.
4. Add an option for minimum request time so that the open request counter delays decrement if a response returns too quickly.
  - Comes close to solving the problem but would require substantial additional complexity, and still requires the user to predict a good value for their workload.
5. Same as 4, but for 5xx responses only. Possibly a more user-friendly implementation, but it makes me uneasy to be mixing application-layer logic here.
6. Go crazy, implement full health-checking of responses and adjust weighting of peers based on (say) percentage of successful requests. Significantly more complexity/risk, and almost certainly more complexity than the ingress-nginx project maintainers are happy to take on.

Out of these I think 1, 2, 3, and 5 are reasonable. I think 1 and 5 have the most predictable behaviour. 2 feels like it doesn't offer a complete enough solution to be worth the trade off in predictability, but I'm certainly happy to discuss it further.

What do you think?

adkafka commented on rootfs/etc/nginx/lua/balancer/leastconn.lua:
```diff
@@ -0,0 +1,112 @@
+
+function _M.balance(self)
+    local peers = self.peers
+    local endpoint = peers[1]
```
How do we behave when `peers` is empty?

adkafka commented on rootfs/etc/nginx/lua/balancer/leastconn.lua:
```diff
@@ -0,0 +1,112 @@
+
+function _M.balance(self)
+    local peers = self.peers
+    local endpoint = peers[1]
```
How do we behave when `peers` is empty?

alowde replied:
I believe that if we reached this line it would cause a Lua exception and Nginx would return an internal server error response, but I'm checking the actual behaviour with some more tests.

adkafka replied:
I haven't come across a test for this yet. Worth adding one?

adkafka commented:
> Out of these I think 1, 2, 3, and 5 are reasonable. I think 1 and 5 have the most predictable behaviour. 2 feels like it doesn't offer a complete enough solution to be worth the trade off in predictability, but I'm certainly happy to discuss it further.

> What do you think?

I think option #1 is best for the scope of this PR. It is the solution of least surprise. I am curious how other load balancers handle these failures though. I know nginx and HA Proxy both have a native least connections LB. Maybe we can look into that a little bit?

Another idea (likely out of scope of this PR) is instead of choosing between the endpoints with the least connections, we place the endpoints in some data structure and choose an endpoint randomly such that the endpoints with fewer connections are more likely to be chosen. This seems like the best balance between converging towards even connection count and limiting the blast radius of a single bad backend. I'm not sure how to implement this yet though...

alowde commented:
> I think option #1 is best for the scope of this PR. It is the solution of least surprise. I am curious how other load balancers handle these failures though. I know nginx and HA Proxy both have a native least connections LB. Maybe we can look into that a little bit?

I've now looked at the documentation for both HAProxy and Nginx Plus and it's interesting to see that neither of them mention this particular issue. I suspect the tighter integration of health-checks and load balancing in both of these means that it's not a major concern. Nginx Plus does have another load balancing mode that integrates least connections with a weighted moving average response time that would be less sensitive to this issue, but it still doesn't actually mitigate it. 
 
> Another idea (likely out of scope of this PR) is instead of choosing between the endpoints with the least connections, we place the endpoints in some data structure and choose an endpoint randomly such that the endpoints with fewer connections are more likely to be chosen. This seems like the best balance between converging towards even connection count and limiting the blast radius of a single bad backend. I'm not sure how to implement this yet though...

Agree that it's not in scope for this PR, but I'm not against implementing a more interesting structure if we can show an advantage for it. If we identify performance issues with the current implementation at scale I would consider assessing peer connection counts asynchronously, and at that point it would be easy to update a second table with an appropriate set of peers to be chosen from randomly.

adkafka commented:
Hey @alowde, I see this PR hasn't had any changes for a while now. What work remains? Anything I can do to help? I would love to use `least_connections` soon :)

alowde commented:
Hey @adkafka, yeah this PR has gotten a little stale - I moved on from the company that was using this and haven't looked at it for a bit. I'd still love to get it fit for merging though, and I'm hopefully able to spend some more time on it now.
Any help would be appreciated. I'll see if I can get it rebased OK this week, and then we'll need to review the existing tests and confirm it's all passing. If I recall correctly some of the test suite had started failing for unrelated features and I didn't track down whether that was because of the code I introduced, updates upstream, or random fluctuations of the universe :).
Feedback/critique would also be welcome on any other aspects of the PR.

tao12345666333 commented:
Thanks! After you rebase this PR, feel free to ping me finish review it

alowde commented:
Just a brief update - I've been actively working on this PR this week but rebasing is taking some work due to overlapping changes with other feature tests added since I started. Currently hoping to have an updated version ready to go within the next week or so, with new cleaner e2e tests.

adkafka commented:
@alowde Great to hear :)

I have been thinking a bit about the conversation thread we had about the downsides to this algorithm: https://github.com/kubernetes/ingress-nginx/pull/9025#issuecomment-1304728049. Specifically, the issue that with a true "least connections" algorithm, all ingress-nginx pods will send traffic to the same pod if that pod just came up. This seems like a potential issue under at least two conditions:
1) An existing deployment is scaled out horizontally, perhaps to handle more load. This will result in only the new pods receiving traffic, and potentially becoming overwhelmed. If we are scaling a deployment from 100 pods to 105, we will have 5 pods handing an incoming request rate that was previously handled by 100 pods (20x the request rate per pod).
2) A pod gets in a bad state for some reason, and is not holding onto new connections. This will result in a complete outage, because all nginx pods send traffic to it.

I believe there is a good solution to this though. I learned recently about "Power of Two Choices" load balancing [1][2]. It is a simpler algorithm, and strikes a good balance between round-robin / random and least connections. Instead of choosing the single backend with the least connections, we would chose two backends at random from the list, and use the one with the least number of connections. The algorithm is simpler to implement than pure least connections, because we don't need to sort through the entire list of backends, we just need to compare two.

I'm bringing this up because for the use case I have in mind, this "Power of Two Choices" + Least Connections load balancing seems idea. It strikes the right balance between spreading connections out over many pods, but still distributing load somewhat evenly. By no means do we have to go this route in this PR, but I want to bring it up for discussion. I'm curious if other use cases would also benefit from this approach, and what other folks think.

[1](https://www.nginx.com/blog/nginx-power-of-two-choices-load-balancing-algorithm/)
[2](https://news.ycombinator.com/item?id=24877341)

reidmeyer commented:
Why was this closed?

alowde commented:
@reidmeyer to rebase this PR I ended up more or less dropping the branch and recreating it, as git rebase wasn't able to produce a clean set of changes. It looks like that caused GitHub to close the PR, but all good now :).

@tao12345666333 this PR is now ready for review, thank you!

mrunelov commented:
Chiming in with a use-case that would prefer a more strict least_connections approach:
Long-running connections like websockets.

I think that could still benefit from some sort of ramp-up though.

reidmeyer commented:
> Chiming in with a use-case that would prefer a more strict least_connections approach: Long-running connections like websockets.
> 
> I think that could still benefit from some sort of ramp-up though.

I have a similar use case: server sent events.

adkafka commented:
> I'd like to leave this PR as is and then raise a new PR to either change the behaviour of the least_connections algorithm or (my preference) add a new algorithm option

That sounds good to me! Perhaps, in a follow-up PR we (I'm happy to do so) can propose a new annotation on the ingress that toggles this behavior (choose two backends randomly, select the one with the fewest connections) on.

adkafka reviewed (commented):
Several of the comment threads are marked "resolved", but the outstanding issue still remains. One example is: https://github.com/kubernetes/ingress-nginx/pull/9025/files#r1200712394. Perhaps a commit was missed?

alowde commented:
@adkafka,

> That sounds good to me! Perhaps, in a follow-up PR we (I'm happy to do so) can propose a new annotation on the ingress that toggles this behavior (choose two backends randomly, select the one with the fewest connections) on.

Sounds good. Please feel free to @ me if/when you do propose a new annotation and I'll be keen to help in any way I can. I think there's an opportunity here to write end-to-end tests that could exercise the load-balancing algorithms with a variety of different traffic patterns, and I'd be very interested to see the outcomes. 

> Several of the comment threads are marked "resolved", but the outstanding issue still remains.

I'm very sorry about this; in the rebase process it appears I was drawing on an old commit to recreate the PR. I've reviewed the conversation and re-resolved the issues raised. Thanks again for the review and your patience.

adkafka commented on test/e2e/framework/deployment.go:
```diff
@@ -256,7 +299,44 @@ http {
+
+// NewAlwaysSlowEchoDeploymentWithOptions creates a new deployment of the always slow echo server with functional options.
+// This server always sleeps for the specified number of milliseconds before responding. NOTE: values for sleepMillis
+// >= 2000 will cause the deployment to fail health checks, causing false positive test failures.
```
Worth adding an `assert` for this? Seems like a time sink on debugging :)

adkafka commented on rootfs/etc/nginx/lua/balancer/leastconn.lua:
```diff
@@ -0,0 +1,108 @@
+            if peer_conn_count  == nil then
+                -- Peer has never been recorded as having connections - add it to the connection
+                -- tracking table and the list of feasible peers
+                endpoints:set(peer_name,0,0)
```
I'm surprised we don't have a default expiry on these entries. Does it make sense to add a long expiry (2 days) by default, just to catch any leaks?

adkafka reviewed (approved):
LGTM! I suggest adding a TTL for all entries in this map, and using that TTL to eventually forget the connection counts for pods. This seems "less surprising" at least to me, but does come at the cost of more memory consumption. Curious what others think.

To make this argument concrete: Let's say we are using a readiness probe on our deployment to prevent a pod from receiving too many connections. Once a predetermined threshold is reached, the pod becomes unready, so the load balancer won't send it anymore traffic. With the implementation as-is, after the pod becomes ready again (connections shed over time), ingress-nginx will think it is a new pod with 0 connections, and send it *all* of the traffic.

ElvinEfendi commented:
A few comments:

- This is not a least connections load balancing, it's load balancing based on the least number of in-flight requests per server. It's a decent approximation of connections, but just wanted to point this out. You might want to rename the algorithm to better represent what it does.
- I'd like to see how this implementation behaves in a k8s cluster with rolling endpoints. How does that compare to EWMA?
- Similar to above, we need some benchmarking numbers.
- I'd suggest using power of two choice technique used in the existing ewma lb algorithm, but instead of using response time as metric, just use the number of active/in-flight requests per server. Moreover, I don't think you need to synchronize request stats across NGINX workers, you can keep it per worker. It's desired in the ewma because the stats are coming from the endpoints in the response headers.

strongjz commented:
A couple things here: 

1. I agree with everything @ElvinEfendi says, i trust his judgement on LUA issues Id like to see what @tao12345666333 thinks as well
2. https://github.com/kubernetes/ingress-nginx/issues/9950
3. https://github.com/kubernetes/ingress-nginx/issues/9949

strongjz commented:
/hold

mrunelov commented:
Regarding what @ElvinEfendi pointed out about in-flight requests, I have a question:
As mentioned I hope to use least_conn for websockets, would those be considered in-flight requests here?

And regarding the algorithm, is there a reason you don't just mimic what already exists in nginx, on a high level?
I'm not well-versed in nginx or ingress-nginx internals mind you :)

ShadowJonathan commented on rootfs/etc/nginx/lua/balancer/leastconn.lua:
```diff
@@ -0,0 +1,108 @@
+    if util.is_blank(upstream) then
+        return
+    end
+    endpoints:incr(upstream,-1,0,0)
```
I have a feeling there is a bug somewhere (most likely with removing/adding endpoints) which either calls this twice, or otherwise does this the "wrong" way;

<img width="1486" alt="image" src="https://github.com/kubernetes/ingress-nginx/assets/22740616/484fc603-a020-4828-8431-69ccd3955938">

These graphs show a massively divergent usage, all similar in increasing amount and such, which leads me to think that this either never reaches "0" after some TTL, or otherwise does not properly "reset" when it sees that it has gone below zero

ShadowJonathan replied:
I know, that code is totally unnecessary, I've changed it around in my fork

(It still seems to exhibit an off-by-one error somewhere after indeterminate amount of time after an indeterminate event, which is why I haven't yet suggested it here)

ShadowJonathan replied:
I've found the off-by-one error, and i've committed it here, if anyone's interested; https://github.com/ShadowJonathan/ingress-nginx/tree/add-least_connections-load-balancing

ShadowJonathan commented:
> I'd like to see how this implementation behaves in a k8s cluster with rolling endpoints. How does that compare to EWMA?

@ElvinEfendi it seems to not particularly like those, at the moment;
<img width="1485" alt="image" src="https://github.com/kubernetes/ingress-nginx/assets/22740616/87a435c4-a38e-42e5-a14a-e32eff66cd3f">

ShadowJonathan reviewed (changes requested):
Btw, `alternative_backends` is not defined, meaning that canary deployments will not properly route and work

strongjz commented:
/assign @tao12345666333

tao12345666333 commented:
Sorry for the long delay. 

I will be back on Tuesday

exherb commented:
what is the status for this pull request? @tao12345666333

em-le-ts commented:
+1

alowde commented:
@exherb this PR is not in a particularly healthy state, sorry. Others have identified specific fixes required and general issues, and you might find @ShadowJonathan 's [fork](https://github.com/ShadowJonathan/ingress-nginx/tree/add-least_connections-load-balancing) useful, but before this code could be merged it would need a fair bit of cleaning up.

I haven't worked at the company which used this code for ~18 months, but if I can get some spare time I'll try to at least bring it up to a usable state.

bashirmindee commented:
any news regarding this ? can we open another PR if this one is becoming stale ?
